uqcsbot/scripts/bgg.py

```diff
@@ -0,0 +1,149 @@
+from requests import get
+from xml.etree.ElementTree import fromstring
+from difflib import SequenceMatcher
+from html import unescape
+
+from typing import Dict, Optional, Any
+
+from uqcsbot import bot, Command
+from uqcsbot.utils.command_utils import UsageSyntaxException, loading_status
+
+
+def get_bgg_id(search_name: str) -> Optional[str]:
+    """
+    returns the bgg id, searching by name
+    """
+    query = get(f"https://www.boardgamegeek.com/xmlapi2/"
+                + f"search?type=boardgame,boardgameexpansion&query={search_name:s}")
+    if query.status_code != 200:
+        return None
+    results = fromstring(query.text)
+    if results.get("total", "0") == "0":
+        return None
+
+    # filters for the closest name match
+    match = {}
+    for i in results:
+        if i.get("id") is None:
+            continue
+        for j in i:
+            if j.tag == "name":
+                match[i.get("id")] = SequenceMatcher(None, search_name, j.get("value")).ratio()
+    return max(match, key=match.get)
+
+
+def get_board_game_parameters(identity: str) -> Optional[dict]:
+    """
+    returns the various parameters of a board game from bgg
+    """
+    query = get(f"https://www.boardgamegeek.com/xmlapi2/thing?stats=1&id={identity:s}")
+    if query.status_code != 200:
+        return None
+    result = fromstring(query.text)[0]
+    parameters: Dict[str, Any] = {}
+    parameters["categories"] = set()
+    parameters["mechanics"] = set()
+    parameters["subranks"] = {}
+    parameters["identity"] = identity
+
+    for i in result:
+        # sets the range of players
+        if i.tag == "poll" and i.attrib.get("name") == "suggested_numplayers":
+            players = set()
+            for j in i:
+                votes = 0
+                for k in j:
+                    votes += (int(k.attrib["numvotes"])
+                              * (-1 if k.attrib["value"] == "Not Recommended" else 1))
+                if votes > 0:
+                    players.add(j.attrib["numplayers"])
+            if players:
+                parameters["min_players"] = min(players)
+                parameters["max_players"] = max(players)
+        # sets the name of the board game
+        elif i.tag == "name" and i.attrib.get("type") == "primary":
+            parameters["name"] = i.attrib["value"]
+        # adds a category
+        elif i.tag == "link" and i.attrib.get("type") == "boardgamecategory":
+            parameters["categories"].add(i.attrib["value"])
+        # adds a mechanic
+        elif i.tag == "link" and i.attrib.get("type") == "boardgamemechanic":
```

Comment:
Pulling things like i.tag, i.attrib, etc. into variables could make this cleaner (or use helper functions).

Reply:
Things pulled into variables.